### PR TITLE
Support catalog.schema.table.column in SQL SELECT and WHERE

### LINKED ISSUE #4872: SQL: WHERE-clauses don't support tables or schemas or catalogs
**Describe the bug**
Columns in `WHERE` clauses cannot be qualified by table or schema.

**To Reproduce**
These tests are added to [`information_schema.slt`](https://github.com/apache/arrow-datafusion/blob/556282a8b6da6cb7d41d8c311211ae49b7ed82a7/datafusion/core/tests/sqllogictests/test_files/information_schema.slt).

This works:

```text
query CCC rowsort
SELECT * from information_schema.tables WHERE table_schema='information_schema';
----
datafusion information_schema columns VIEW
datafusion information_schema df_settings VIEW
datafusion information_schema tables VIEW
datafusion information_schema views VIEW
```

---

This fails:

```text
query CCC rowsort
SELECT * from information_schema.tables WHERE tables.table_schema='information_schema';
----
datafusion information_schema columns VIEW
datafusion information_schema df_settings VIEW
datafusion information_schema tables VIEW
datafusion information_schema views VIEW
```

with

```text
Error: SqlLogicTest(query failed: DataFusion error: Schema error: No field named 'tables'.'table_schema'. Valid fields are 'information_schema.tables'.'table_catalog', 'information_schema.tables'.'table_schema', 'information_schema.tables'.'table_name', 'information_schema.tables'.'table_type'.
````

---

This fails:

```text
query CCC rowsort
SELECT * from information_schema.tables WHERE information_schema.tables.table_schema='information_schema';
----
datafusion information_schema columns VIEW
datafusion information_schema df_settings VIEW
datafusion information_schema tables VIEW
datafusion information_schema views VIEW
```

with

```text
Error: SqlLogicTest(query failed: DataFusion error: Error during planning: Unsupported compound identifier 'Full { catalog: "information_schema", schema: "tables", table: "table_schema" }'
```

---

This fails:

```text
query CCC rowsort
SELECT * from information_schema.tables WHERE datafusion.information_schema.tables.table_schema='information_schema';
----
datafusion information_schema columns VIEW
datafusion information_schema df_settings VIEW
datafusion information_schema tables VIEW
datafusion information_schema views VIEW
```

with

```text
Error: SqlLogicTest(query failed: DataFusion error: Error during planning: Unsupported compound identifier '[Ident { value: "datafusion", quote_style: None }, Ident { value: "information_schema", quote_style: None }, Ident { value: "tables", quote_style: None }, Ident { value: "table_schema", quote_style: None }]'
```

**Expected behavior**
All tests pass.

**Additional context**
Tested w/ commit https://github.com/apache/arrow-datafusion/commit/556282a8b6da6cb7d41d8c311211ae49b7ed82a7 .

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #4872

# Rationale for this change

Per original issue, be able to support SQL such as

```sql
select datafusion.public.test.a from test where datafusion.public.test.b = 1
```

This required a larger refactoring specifically around `Column`, to change `relation` from an `Option<String>` to `Option<OwnedTableReference>`, since otherwise can only support a single level of identifier for the relation of a column (unless do some weird stuff like serializing/deserializing into that single String)

# What changes are included in this PR?

Change `Column` to have `Option<OwnedTableReference>` for `relation` field. Accompanying fixes to rest of codebase for this.

Also fix for the original issue, where in parsing sql compound identifiers can properly identify identifiers which are comprised like `catalog.schema.table.column` which previously wasn't supported.

# Are these changes tested?

Unit test added & fixed, also added test to sqllogictest from the original issue

# Are there any user-facing changes?

Yes, breaking changes

### CODE REVIEW

File: datafusion/common/src/dfschema.rs
Comment (Jefffrey) [inline datafusion/common/src/dfschema.rs:591]: originally had this way to try ease the breaking change, since would still support Option<&str>, but can see will cause pain if trying to pass in just None itself. maybe its better to just have this as Option<OwnedTableReference> and have the callers call into() (or maintain old method and create a new one?)

File: datafusion/common/src/dfschema.rs
Comment (Jefffrey) [inline datafusion/common/src/dfschema.rs:690]: this was a separate change i decided to add, mainly since i believe having quoted identifiers allows for more robust messages as itll properly account for any special characters within and allows users to clearly delineate identifiers in that case

e.g. in old way a user could have an ident like `t1'.'t2` which would show in error as `'t1'.'t2'` even though it is a single ident, it looks like two

File: datafusion/common/src/error.rs
Comment (Jefffrey) [inline datafusion/common/src/error.rs:134]: otherwise clippy lint about large Err variant appeared

File: datafusion/sql/src/expr/identifier.rs
Comment (Jefffrey) [inline datafusion/sql/src/expr/identifier.rs:119]: main  fix for original issue is in here

File: datafusion/common/src/column.rs
Comment (Jefffrey) [inline datafusion/common/src/column.rs:40]: by having this as Into<OwnedTableReference>, means if passing in a string it will be parsed. this has the side-effect of if passing in just a string like `C1` for the column name, itll be normalized to `c1` (wheres previously it wasn't, since was passed straight through)

if this breaking behaviour is too large, can keep the old method (and just create an `OwnedTableReference::Bare` from it) and have a new method which accepts `Option<OwnedTableReference>` for the relation

see documentation change that was needed due to `col(...)` behaviour changing for reference

**Review by Jefffrey — COMMENTED** [review]
I apologize for such a size PR, especially in regards to the amount of breaking API changes involved, and am happy for any feedback, especially if you see a better way to achieve the original issue without causing as much breakage

**Comment by alamb** [comment]
I plan to review this PR carefully tomorrow. I am sorry for the delay

File: datafusion/common/src/table_reference.rs
Comment (alamb) [inline datafusion/common/src/table_reference.rs:399]: 👍

File: datafusion/expr/src/logical_plan/builder.rs
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:549]: I tried playing around with the signature of `field_with_qualified_name` and I couldn't find a way to make `lr` still work. 🤷

File: datafusion/optimizer/src/push_down_projection.rs
Comment (alamb) [inline datafusion/optimizer/src/push_down_projection.rs:497]: maybe worth a follow on ticket?

File: docs/source/user-guide/example-usage.md
Comment (alamb) [inline docs/source/user-guide/example-usage.md:123]: This is probably a pretty major change for people (that `col()` will now parse / normalize the column names according to SQL semantics).

However, it does have the nice side benefit that the handling of `.` is no longer treated differently in `col()`

What do you think about possibly 
1. changing `col` so that it forms a unqualified reference always (aka does not respect `.`)
2. adding some new function like `ident()` or `parse_col()` that treats column like a sql identifier and correctly splits based on `.` etc?

File: datafusion/sql/src/expr/identifier.rs
Comment (alamb) [reply datafusion/sql/src/expr/identifier.rs:119]: Could you write some more unit tests for these cases (the comments are great, but it would be better if they were tests so they can't get out of sync as easily)

**Review by alamb — COMMENTED** [review]
First of all, thank you very much @Jefffrey  -- I think this PR is very well coded and structured. Very nice. I very much apologize for the delay in reviewing

As one measure of how impactful this change might be, I ran the code on this branch against the IOx test suite (see https://github.com/influxdata/influxdb_iox/pull/7077) and it passed with minor changes. Nice work

Here is what I think is needed prior to merge:
1. Some additional testing for name resolution in `identifier.rs` (basically turn the comments into tests)
2. Some discussion with the broader community about how to handle `col("A")` being normalized to`col("a")` -- I think this may be very surprising to some users and I offered an alternate suggestion (`ident`, that is still a breaking API change but may be less confusing)

cc @houqp  who did some of the original work on multi-part identifiers

cc @liukun4515  @andygrove @avantgardnerio @thinkharderdev @jackwener @ozankabak  in case they have thoughts about this API change or the structure in genera

File: datafusion/common/src/dfschema.rs
Comment (Jefffrey) [reply datafusion/common/src/dfschema.rs:591]: An issue with this approach is that there are still use cases for an Option<...> parameter, such as when the caller already possesses an `Option<OwnedTableReference>` (such as the `relation` field of `Column`)

This would force them to destructure that Option manually to call the correct version, `new(...)` or `new_unqualified(...)`, which makes it more verbose for the caller

I suppose could add yet another version of new... I'm not sure I like that approach either though

File: datafusion/common/src/dfschema.rs
Comment (Jefffrey) [reply datafusion/common/src/dfschema.rs:690]: Yeah this did come to mind, but:

- Either have to store alongside the name that it was normalized
- Or each time quoting string have to do a pass over it to check for special characters/uppercase characters

1st one I don't really like as have to add more to TableReference (and related structs)

2nd one would be simpler to implement, maybe I'm overthinking the impact of the overhead 😅

File: datafusion/common/src/column.rs
Comment (Jefffrey) [reply datafusion/common/src/column.rs:40]: Sure thing

File: datafusion/expr/src/logical_plan/builder.rs
Comment (Jefffrey) [reply datafusion/expr/src/logical_plan/builder.rs:549]: Yeah was thinking maybe having something like `Into<TableReference>` or the like.

Related, but I feel TableReference and its related structs OwnedTableReference and ResolvedTableReference got a bit out of hand, since they seem all very similar and have conversions between them, but still are quite separate. I think I initially had `field_with_qualified_name` accept a reference to an OwnedTableReference but changed it to TableReference as that wouldn't require owning a bunch of strings if you just wanted to pass in a reference, but can see it doesn't really result in a nice API imo

Maybe need refactoring around the TableReference structs to try make them more interchangeable, via a common trait possibly

File: docs/source/user-guide/example-usage.md
Comment (Jefffrey) [reply docs/source/user-guide/example-usage.md:123]: Yeah that would have the benefit of being less breaking and being a bit more explicit in what is happening. `parse_col` would make it more explicit though I'm not really a fan of that name, feels a bit clunky.

For what its worth Spark does seem to be able to parse the input to their `col` function, instead of taking it wholly unqualified I believe

File: datafusion/expr/src/logical_plan/builder.rs
Comment (alamb) [reply datafusion/expr/src/logical_plan/builder.rs:616]: > Maybe need refactoring around the TableReference structs to try make them more interchangeable, via a common trait possibly

I wonder if we could maybe change TableReference to use `Cow` somehow rather than separate variants 🤔

**Comment by alamb** [comment]
So in my opinion this PR, while a non trivial breaking change, results in a better outcome -- however, before we merge it let's solict additional community feedback. 

I posted a note in the slack channel and the dev list https://lists.apache.org/thread/m42rk0sbh88bnbgnx83wshhm97wrxjf7

**Comment by ozankabak** [comment]
I agree that this is a good change and the API change is OK with us. However, honestly I can not confidently tell that it won't mess with others' usage patterns. I think waiting for some more community feedback is a good idea.

File: datafusion/expr/src/logical_plan/builder.rs
Comment (Jefffrey) [reply datafusion/expr/src/logical_plan/builder.rs:616]: How do you mean?

File: datafusion/common/src/dfschema.rs
Comment (Jefffrey) [reply datafusion/common/src/dfschema.rs:635]: In the end I opted to leave `new(...)` as is with the `Option<...>` parameter, and introduced a new `new_qualified(...)` to allow specifying without qualifier, without needing to turbofish for a None case

File: datafusion/sql/src/expr/identifier.rs
Comment (Jefffrey) [reply datafusion/sql/src/expr/identifier.rs]: Refactored and added unit tests

File: datafusion/common/src/column.rs
Comment (Jefffrey) [reply datafusion/common/src/column.rs:45]: Doc updated

**Comment by avantgardnerio** [comment]
At a high-level, this seems necessary but painful :/

File: datafusion/expr/src/logical_plan/builder.rs
Comment (alamb) [reply datafusion/expr/src/logical_plan/builder.rs:612]: I was thinking about removing OwnedTableReference -- I started to sketch out what this might look like in https://github.com/apache/arrow-datafusion/pull/5508

Working with the structures then I think becomes less cumbersome

File: docs/source/user-guide/example-usage.md
Comment (alamb) [reply docs/source/user-guide/example-usage.md:123]: How about we simply add an `ident()` function that doesn't parse at all? I can help do that

**Review by alamb — COMMENTED** [review]
I went through this PR again I think we should merge it because even though it is disruptive, it fixes a bug and further more a long standing source of bugs (from the inconsistent behavior of `col()`)

I recommend as next steps:
1. Add an `ident()` or `name()` function that allows users to refer to columns like `A` without having to use \"A\"
2. Possibly reconsider OwnedTableReference. Here is a proposal of how to handle `OwnedTableReference`s: https://github.com/apache/arrow-datafusion/pull/5508/files

 I think the `ident` / `name` is important before merge. The OwnedTableReference thing we can do as a follow on PR

THanks again @Jefffrey  for pushing this forward

File: docs/source/user-guide/example-usage.md
Comment (Jefffrey) [reply docs/source/user-guide/example-usage.md]: I've added the `ident()` expr function and updated the `user-guide/example-usage.md` doc

Let me know if there are other places I need to change/add docs for

**Comment by Jefffrey** [comment]
I've done the changes for `OwnedTableReference` per your suggestion @alamb

**Review by alamb — APPROVED** [review]
All in all this is really nice work @Jefffrey  -- it demonstrates a wonderful knowledge of the DataFusion codebase and I think cleans up some long standing techdebt (@houqp would be proud)

Here is what I think we should do with this PR given it has such a significant API change (at least I think so), unless anyone objects:

1. Wait for the next datafusion release (should be cut on this Friday, March 10)
2. Merge the PR first thing

This will give us the maximal time to let the dust settle, and a few bleeding edge adopters to pick it up (aka I will integrate it into IOx, and fix any bugs / rough edges we might find during that process).

cc @andygrove @jdye64 @Dandandan   (this PR this will mean that anyone that uses `col("FoorBar")` it will be the same as `col(foobar)` which is not the same today (the colum reference is not normalized).

@Jefffrey has added the `indent()` function that has similar behavior to the old one

File: datafusion/sql/src/expr/identifier.rs
Comment (Jefffrey) [reply datafusion/sql/src/expr/identifier.rs:377]: I made it `Internal` error since `form_identifier(...)` is a private function that is only used by `sql_compound_identifier_to_expr(...)` which already checks that it doesn't pass an empty ident list in (hence calls unwrap on it).

So in normal usage this error can't ever occur. I made it return an error type just in case in future the method is used elsewhere. I'll update the comments a bit to reflect this

File: datafusion/common/src/dfschema.rs
Comment (Jefffrey) [reply datafusion/common/src/dfschema.rs:744]: raised ticket: https://github.com/apache/arrow-datafusion/issues/5523

File: datafusion/expr/src/expr_fn.rs
Comment (Jefffrey) [reply datafusion/expr/src/expr_fn.rs:50]: Doc updated

File: datafusion/expr/src/expr_fn.rs
Comment (Jefffrey) [reply datafusion/expr/src/expr_fn.rs]: Updated doc to have example

File: datafusion/core/src/execution/context.rs
Comment (Jefffrey) [reply datafusion/core/src/execution/context.rs:332]: Yeah this looks good, as it looks better without the clone, however I think `to_owned_reference(...)` will clone the underlying strings anyway:

https://github.com/Jefffrey/arrow-datafusion/blob/57ce66c24858cedac4f2c56194ea72f84d7be2f5/datafusion/common/src/table_reference.rs#L192-L212

**Comment by alamb** [comment]
I plan to merge this PR tomorrow unless anyone else has comments or wants more time to reviwe

**Comment by alamb** [comment]
Thanks again @Jefffrey  -- this is pretty epic. It may be disruptive but I think the new consistency will be worth it

**Pull request merged**
